src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -408,15 +379,17 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
                 methodSignatureOpt: method,
                 returnTypes: null,
                 initialState: null,
-                analyzedNullabilityMapOpt: null);
+                analyzedNullabilityMapOpt: null,
+                checkpointMapOpt: null);
         }
 
         internal static BoundNode AnalyzeAndRewrite(
             CSharpCompilation compilation,
             Symbol symbol,
             BoundNode node,
             Binder binder,
-            DiagnosticBag diagnostics)
+            DiagnosticBag diagnostics,
+            Dictionary<BoundNode, Checkpoint> checkpointMapOpt)
```

Comment:
>Dictionary<BoundNode, Checkpoint> checkpointMapOpt [](start = 12, length = 50)

Could this be returned as an `ImmutableDictionary<,>`?

Reply:
I expect this method to be used both by the speculative semantic path and non-speculative path, and the speculative path won't need to actually do the checkpointing. I could change this to be a bool flag about whether to do the checkpointing and return null in that case if you'd prefer.

---
In reply to: [286218983](https://github.com/dotnet/roslyn/pull/35850#discussion_r286218983) [](ancestors = 286218983)

Reply:
Changing to `bool createMap, out ImmutableDictionary<,> map` sounds reasonable.

---
In reply to: [286219930](https://github.com/dotnet/roslyn/pull/35850#discussion_r286219930) [](ancestors = 286219930,286218983)